### ISSUE TITLE
Ulauncher not displaying applications if launched manually

### ISSUE THREAD

**Issue by KakolIsSomewhatGay** [Ulauncher/Ulauncher#241]
*(If launched manually)* Ulauncher will not display applications in results even though it is able to access `~/.local/share/applications` and `/usr/share/applications`. Is this intended?

**Steps to reproduce:**
Install using ppa (stable)
Start the application via the Dash

**Ulauncher version**: 4.0.8.r1
**Linux distro**: Ubuntu
**Desktop environment**: Unity

[Term Gist](https://gist.github.com/KakolIsSomewhatGay/1b13b6d92024032ce8ae5391e02c573a)
 #~~**Future Persons:** Ulauncher appears to not cache application data, so if you have alot of `.desktop` links it may take a while to index it all. #154~~
**Note:** If started manually Ulauncher will have the `Launch at Login` option greyed out

**Comment by KakolIsSomewhatGay** [comment]
After fiddling with CCSM to get unity to default to factory settings Ulauncher is now displaying applications. Don't exactly know what I did that got it working though

**Comment by gornostal** [comment]
@KakolIsSomewhatGay 

App indexing takes some time. I have an issue for that (#154).
If you wait for a minute or two, does it still show no apps?

Option is greyed out due to the same reason -- Ulauncher has to find itself among other apps to be able to add its .desktop file to autostart.

I understand this is not the best UX. I'm planning to fix that.

**Comment by KakolIsSomewhatGay** [comment]
@gornostal

Apologises for the slow response, I don't check my github very often.
Ulauncher does eventually index all applications on my system however it takes quite a while, I'd say around 10-ish minutes. Even once indexing is finished the autostart checkbox is still greyed out.

Although index caching could help the issue I think it would be more future-proof to make the indexing faster, results not showing up for first-time users is just going to end up confusing them.

**Comment by gornostal** [comment]
In order to make indexing faster you can do the following:

1. Open `~/.cache/ulauncher_cache/last.log` and look for errors like this:
    ```
    2018-05-31 22:39:08,127 | INFO | ulauncher.util.desktop.reader: read_desktop_file() | Could not read "/usr/share/applications/evolution-calendar.desktop": constructor returned NULL
    ```
    You'll probably see lot's of them with the same directory prefix.
2. Add directory prefix with errored desktop files to Blacklisted App Dirs in Ulauncher preferences.
    You don't want to add `/usr/share/applications` to that list, but something like `/usr/share/fk5` will probably have desktop files that are no useful and take up most of the time to be indexed.

As for the hotkey, I'm not sure what you mean by "Ulauncher will block Hotkey shortcuts". Could you explain a bit more?

**Comment by KakolIsSomewhatGay** [comment]
@gornostal Say for example you have the (Super|Win) key set as your hotkey via Ulauncher, attempting to press the key will instantly open Ulauncher Search. Perfect!

However this has the side effect of stopping key combinations such as `[Hotkey] + D` since Ulauncher opens when they key is pressed instead of when it is released, for multi-key combos such as `Ctrl + Space` this probably wont be an issue but for single keys such as Super it can.]

I've added the failing folders to the blacklist although not many and I'll get back to you if I see an improvement in times

**Comment by gornostal** [comment]
Thanks for the clarifications.
Unfortunately, there is no solution to your issue with the hotkey.
And I wasn't planning to add option to trigger Ulauncher on key release.

Have you been able to resolve the issue with slow app indexing by following the steps above?

**Comment by KakolIsSomewhatGay** [comment]
Sadly no, Ulauncher still takes a long time to index. Maybe a path whitelist could make indexing faster?

**Comment by KakolIsSomewhatGay** [comment]
@gornostal 
`2018-06-13 03:27:32,985 | INFO | ulauncher.search.apps.app_watcher: start() | Scanned desktop dirs in 2692.42 seconds. Indexed 187 applications`
oof

**Comment by gornostal** [comment]
40 minutes is quite a long wait for apps to be indexed.
Could you attach `~/.cache/ulauncher_cache/last.log` please?

BTW, what do you think may be a reason for that? Do you have slow CPU or HDD?

**Comment by KakolIsSomewhatGay** [comment]
@gornostal Not exactly, I've never had any performance issues with my machine. [last.logGist](https://gist.github.com/KakolIsSomewhatGay/bdcb645ab03672fbc46e007289fb0a2a)

**Comment by gornostal** [comment]
Thanks for the logs.
Strangely enough, there is nothing in the logs that would explain that long indexing time.

**Comment by gornostal** [comment]
If you'd like to spend some more time debugging, here's what you can do.
This issue may be due to a large number of files in directories that Ulauncher uses for searching apps.
To verify that, you can run `python` from CLI and then enter this:

```
import os; from xdg.BaseDirectory import xdg_data_dirs; filter(os.path.exists, xdg_data_dirs)
```

You'll get a list of directories. Then you can run this command to count files in each directory:

```
find <div> -type f | wc -l
```

For the reference, I have around 94K files in `/usr/share` and 27K in `~/.local/share`

**Comment by KakolIsSomewhatGay** [comment]
For `/usr/share` I've got 112K, and 383K for `~/.local/share`. Couldn't a folder whitelist help with users that have a a lot of files?

**Comment by gornostal** [comment]
It could, but it will also add either complexity to the code (which I don't want) or complexity to Ulauncher Preferences UI (which users may not want).

I'm not sure what's the best solution for this issue. Need to think about it...

**Comment by KakolIsSomewhatGay** [comment]
A `Whitelisted App Dirs` list could be added with a toggle to enable it, I'd suggest it and `Blacklisted App Dirs` should be collapsible though

**Comment by gornostal** [comment]
Yeah, I understand that. But that means more code to support. Instead, I want to reduce amount of code

**Comment by KakolIsSomewhatGay** [comment]
Blacklist and Whitelist Filters could be moved to an extension and allow it to be added by users if they want that customisation

**Comment by gornostal** [comment]
Hi @KakolIsSomewhatGay 
I've just released Ulauncher 4.1.0 which includes app caching (list of the indexed apps will be cached after the first launch, so on the consecutive launches you'll instantly see the apps). 
It's not an ideal solution to your issue, but that definitely should help.

I'm going to close this issue, because there is not much that can be done about it.
